src/miner.cpp

```diff
@@ -641,10 +641,13 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddre
         // Fill in header
         pblock->hashPrevBlock  = pindexPrev->GetBlockHash();
         if (chainparams.GetConsensus().NetworkUpgradeActive(nHeight, Consensus::UPGRADE_NU5)) {
-            // hashBlockCommitments depends on the block transactions, so we have to
-            // update it whenever the coinbase transaction changes. Leave it unset here,
-            // like hashMerkleRoot, and instead cache what we will need to calculate it.
             pblocktemplate->hashChainHistoryRoot = view.GetHistoryRoot(prevConsensusBranchId);
+            // hashBlockCommitments depends on the block transactions, so we have to
+            // update it whenever the coinbase transaction changes. But set it here so that
+            // it's available to the getblocktemplate RPC.
+            pblock->hashBlockCommitments = DeriveBlockCommitmentsHash(
+                pblocktemplate->hashChainHistoryRoot,
+                pblock->BuildAuthDataMerkleTree());
```

Comment:
This actually exposes a more serious issue with the data we are returning from `getblocktemplate`. The reason why the code was the way it was before this PR, is that `hashBlockCommitments` depends on the exact transactions in the block, including the coinbase tx, just like `hashMerkleRoot` does. When `getblocktemplate` returns data, it _does not_ return `hashMerkleRoot`; instead, it returns the selected transactions, and the suggested coinbase transaction, but leaves it entirely up to the caller to compute `hashMerkleRoot`. (In Bitcoin this would happen regularly because they shove extra nonce data into the coinbase, but because we fixed that by making the header nonce field larger, Zcash miners don't normally need to alter the coinbase tx.)

The thing is, `hashBlockCommitments` actually needs to be treated by callers the same way as `hashMerkleRoot`: if they change any txs, they need to recompute `hashMerkleRoot`, _and_ recompute `hashAuthDataRoot` (from ZIP 244), and then use the latter to recompute `hashBlockCommitments`. So while we _can_ calculate the expected `hashBlockCommitments` for the caller, doing so doesn't actually "work" the way the rest of `getblocktemplate` works. However, we don't currently supply enough information for the caller to be able to recompute `hashBlockCommitments` without also implementing all the history tree logic.

I think we should make the following changes:
- Remove `blockcommitmentshash` from the `getblocktemplate` output.
  - It's only been used on testnet thus far (and clearly was giving incorrect output), so I have no issue with making that backwards-incompatible change.
- Un-deprecate `lightclientroothash`, and have it continue to return the light client root hash even after NU5.
  - This would technically not be a breaking change from its current semantic meaning on mainnet, but might be breaking in how mining pools were assuming it could be used (shove it into the header).
- Document in the `getblocktemplate` help text that callers need to compute both `hashMerkleRoot` (as usual) and `hashBlockCommitments` (per ZIP 244) using whatever txs end up in their block.

Reply:
Hm. Asking the callers to implement ZIP 244 seems like a bit of a lift. Could we instead do something where we add an optional list of txids as an argument to `getblocktemplate`, include those transactions, and then compute `hashBlockCommitments` for them?

Reply:
> However, we don't currently supply enough information for the caller to be able to recompute hashBlockCommitments without also implementing all the history tree logic.

This seems like an unreasonable ask for them to do without help.

Reply:
> Could we instead do something where we add an optional list of txids as an argument to `getblocktemplate`

I don't think this would be useful, because I don't think miners typically compute transactions (not in the mempool) to add to their blocks. Besides, if the caller submits a txid (as a `getblocktemplate` argument) that isn't in the mempool, that transaction will be unknown to `zcashd`, so `getblocktemplate` can't include it. It's like you would have to provide the full transaction, not just its txid.

I think what we suspect is actually the case -- miners just use the exact set of transactions that `getblocktemplate` returns. I like the idea of providing all the information needed for this "special" (not so special, very common) case; the miner always has the option to recompute these values if it wants to modify the transaction list. Documentation, of course, is key here.

Reply:
The conclusion we came to from the meeting this morning was:
1) Both `hashBlockCommitments` and `hashMerkleRoot` should be provided in a portion of the block template JSON such that, if the miner accepts the block & coinbase transaction structure as proposed, will just work.
2) It is correct to remove lightclientroothash, because there was a potential bug related to us providing that value anyway.
3) It should be documented that if a miner wishes to build the block themselves, they'll need to correctly implement ZIP 244 for computing the values of these two commitments.

Reply:
More specifically, the `mutate` field of the RPC output specifies the fields that the caller can mutate in the template. Of these:
- Mutating `time` has no effect on other header fields.
- Mutating `transactions` means the caller needs to recompute `hashMerkleRoot` (which they need to anyway), and `hashAuthDataRoot` (thus also recomputing `hashBlockCommitments`).
- Mutating `prevblock` means the caller needs to recompute `hashLightClientRoot` (thus also recomputing `hashBlockCommitments`).

So the "special" portion of the output should contain the three base values (`hashMerkleRoot`, `hashAuthDataRoot`, and `hashLightClientRoot`), for which we are providing correct values assuming no mutations. It should also contain the derived value `hashBlockCommitments` for convenience.

If we want consistency with the `getblock` output, the new `hashMerkleRoot` field (which until now was not present) should be named `merkleroot`. We could also make the names of the other fields `blockcommitmentshash` and `lightclientroot` to similarly match that naming scheme (where the `hash` suffix is only used for block hashes, and `root` is the suffix for tree roots).

Reply:
There's a possibility that some mainnet mining pools have already switched to using `blockcommitmentshash` for pre-NU5 usage, so we can't actually remove it. However, we _can_ decree that mining pools will all need to move to the new fields in the special section (`defaultroots`?). So what we can do is deprecate `blockcommitmentshash`, leave `lightclientroothash` deprecated, and then once NU5 activates, we can drop all the deprecated fields (since we know no one will be using them anymore).

Reply:
> So the "special" portion of the output should contain the three base values (`hashMerkleRoot`, `hashAuthDataRoot`, and `hashLightClientRoot`)

Two questions:

- Why does `hashAuthDataRoot` need to be included in the new `defaultroots` section? It's not part of the block header, and has to be recomputed if the miner changes the transactions (correct?). How would that value be useful to the miner?
- Should the `defaultroots` section contain a `lightclientroothash` entry or did you mean to say `blockcommitmentshash`?

Reply:
`hashAuthDataRoot` is part of the block header (it is part of what needs to be concatenated together to produce `hashBlockCommitments`). If the miner doesn't change the transactions, they can just concatenate `hashLightClientRoot || hashAuthDataRoot || [0u8; 32]` to get the value of `hashBlockCommitments`

Reply:
- The value of including `hashAuthDataRoot` is that if the miner leaves the transaction set as-is, but alters `hashPrevBlock`, then they will need to recompute `hashLightClientRoot`, and then `hashBlockCommitments`, but can reuse `hashAuthDataRoot`. If we only included `hashBlockCommitments`, then `hashAuthDataRoot` would need to be recomputed even if the transaction set didn't change.
- Yes it needs to contain that entry (as we just discussed in our pairing), for the complementary reason to above: if the miner alters the transaction set but not `hashPrevBlock`, they can reuse `hashLightClientRoot`.